File: src/main/java/com/example/Toros/model/Tournament.java
package com.example.Toros.model;

import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Entity
@Table(name = "tournaments")
@Data
public class Tournament {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;
    private String name;
    private boolean isActive; // Indicates if the tournament is still ongoing
    private String status; // "LIVE", "COMPLETED", "UPCOMING"
    private int numCourts;
    private int gamesPerTeam;
    private boolean tiered;
    private Integer breakDuration; // In minutes
    private Integer matchDuration; // In minutes
    private LocalDateTime startTime;
    private LocalDateTime endTime;
    private String location; // Location of the tournament
    private String organizer; // Name of the organizer
    private String contactInfo; // Contact information for the organizer
    private String tournamentType; // e.g., "Singles", "Doubles", "Mixed"
    private String scoringSystem; // e.g., "Rally Scoring", "Traditional Scoring"
    private String rules; // e.g., "USAPA Rules", "Custom Rules"
    private String prizeDistribution; // e.g., "1st: $500, 2nd: $300, 3rd: $200"
    private String format; // e.g., "Round Robin", "Single Elimination", "Double Elimination"
    private String ageGroup; // e.g., "18+", "35+", "50+"
    private String skillLevel; // e.g., "Beginner", "Intermediate", "Advanced"
    private String auth0Id; // Auth0 ID for the organizer, if applicable
    private String accessCode; // Access code for referees to join

    @ElementCollection
    private List<String> setupProperties; // Example: ["Match Duration: 15min", "Double Elimination: true"]

    @ElementCollection
    private Map<String, String> setupPropertiesMap; // PostgreSQL handles Map<String, String> better as
                                                    // ElementCollection

    // Now storing only IDs rather than full objects
    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "tournament_id")
    private List<Team> teams = new ArrayList<>();

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "tournament_id")
    private List<Match> matches = new ArrayList<>();

    @ElementCollection
    private List<String> finalPlacements; // Final ranking after tournament completion

    @ElementCollection
    private List<String> players;
}

File: Toros.Backend/Toros.Backend/Models/Tournament.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Toros.Backend.Models;

[Table("tournaments")]
public class Tournament
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public string Id { get; set; } = Guid.NewGuid().ToString();

    [Required]
    public string Name { get; set; } = string.Empty;

    public string Status { get; set; } = "DRAFT"; // DRAFT, LIVE, COMPLETED

    public string Location { get; set; } = string.Empty;
    public string Date { get; set; } = string.Empty;
    public string Time { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;

    public int MaxTeams { get; set; }
    public int NumberOfCourts { get; set; }

    public string OrganizerName { get; set; } = string.Empty;
    public string OrganizerContact { get; set; } = string.Empty;

    public string Type { get; set; } = "Round Robin"; // Round Robin, Single Elim, Double Elim
    public string AgeGroup { get; set; } = string.Empty;
    public string SkillLevel { get; set; } = string.Empty;

    public string? Auth0Id { get; set; }
    public string? AccessCode { get; set; }

    // Navigation properties
    public List<Team> Teams { get; set; } = new();
    public List<Match> Matches { get; set; } = new();

    // Map/List properties - stored as JSONB in Postgres
    [Column(TypeName = "jsonb")]
    public List<string> SetupProperties { get; set; } = new();

    [Column(TypeName = "jsonb")]
    public Dictionary<string, string> SetupPropertiesMap { get; set; } = new();

    [Column(TypeName = "jsonb")]
    public List<string> FinalPlacements { get; set; } = new();

    public List<string> RegisteredPlayerIds { get; set; } = new();
}
